### PR TITLE
Describe an `Enum`'s options by its member docstrings when it mixes in `UseEnumMemberDocstrings`

### PR DESCRIPTION
**Pull request by DouweM**
An `Enum` that mixes in `UseEnumMemberDocstrings` describes each of its options to the model by the docstring written under that member. Pydantic reads a docstring under a model field (`use_attribute_docstrings`) but not one under an enum member, so until now the only way to explain an option was to write the explanations into the field's description by hand.

```python
from enum import Enum

from pydantic_ai import UseEnumMemberDocstrings


class Priority(UseEnumMemberDocstrings, str, Enum):
    """How urgent the ticket is."""

    low = 'low'
    """Can wait a week."""
    high = 'high'
    """Needs attention today."""
```

renders in tool and output schemas as

```json
{"anyOf": [{"const": "low", "description": "Can wait a week."}, {"const": "high", "description": "Needs attention today."}], "description": "How urgent the ticket is.", "type": "string"}
```

which is the JSON Schema way to describe single values.

### Opt-in, and why it's a base class

Reading the docstrings unconditionally would have changed the schema of every enum every Pydantic AI user already has, so it is opt-in. **An enum that does not mix `UseEnumMemberDocstrings` in is described exactly as Pydantic itself describes it** — byte-identical to today's shape — so upgrading moves nothing for anyone who doesn't ask for it. `tests/test_tools.py::test_an_enum_that_does_not_mix_it_in_is_described_exactly_as_pydantic_describes_it` pins that against Pydantic's own generator rather than against a snapshot alone.

The opt-in is a base class rather than a config flag because an `Enum` has nowhere else to put one: it has no `model_config`, and it cannot carry a plain class attribute either, since any assigned value becomes a member whether it is annotated or not. Gating on the enclosing model's `use_attribute_docstrings` was tried and doesn't work: that config is pushed while the *core* schema is built and nothing pushes it while the JSON schema is generated, so an enum reached from a tool's parameters never sees it even though `_function_schema` sets it — a rule that would have held in three places out of four, with nothing to tell the user which one they were in.

On the name: `UseEnumMemberDocstrings` rather than `UseAttributeDocstrings` because "member" is the enum-specific word — Python says `__members__` and `enum.nonmember` — so it can't be misread as turning on attribute docstrings for a model's fields, which is what Pydantic's `use_attribute_docstrings` covers. The `Enum` prefix is deliberately redundant at the usage site (`class Priority(UseEnumMemberDocstrings, str, Enum)`) because the name is also read in isolation, in docs, in a changelog and in autocomplete, where "member" alone would leave the scope ambiguous. It is nonetheless the enum counterpart to `model_config = ConfigDict(use_attribute_docstrings=True)`, and the docs say so.

The class lives in `_json_schema.py` — it applies to output types as much as to tool parameters — and is exported as `from pydantic_ai import UseEnumMemberDocstrings`.

### Edge cases

Members without a docstring keep a bare `const`, and every `Literal` renders as before. A docstring under an alias (`urgent = 'high'` beside `high = 'high'`) describes the option it was written for, since the schema only ever names the canonical member. An enum with a `None` member keeps the plain `enum` list, because `{'const': None}` reads as "no const" to anything that looks the key up with a default and the option would silently lose its constraint. The docstrings are read from source with `ast`, so an opted-in enum built by the functional API, or one defined in the REPL, stays a plain `enum`.

Split out of #8450, where `TypeSafeModel` maps a rubric's levels to Jev's score primitive by these docstrings, because this changes the schema a provider receives and deserves its own review and release. Reviewer's request on #8450: recorded tests that a real transformer accepts the shape on the wire, not only a shape snapshot.

### Verification

- `tests/test_tools.py::test_mixing_in_use_enum_member_docstrings_is_what_describes_the_options` and `::test_an_enum_that_does_not_mix_it_in_is_described_exactly_as_pydantic_describes_it` are two enums with identical bodies differing only in the mix-in, so the opt-in is the only variable.
- `::test_enum_member_docstrings_describe_options` pins the schema shape for documented, undocumented and source-less members; `::test_an_enum_alias_describes_the_option_it_was_written_for` and `::test_a_none_member_keeps_the_enum_in_its_plain_form` pin the edge cases.
- Recorded wire tests for OpenAI, Anthropic, Google and Mistral send a tool with a documented enum and assert the call comes back with one of its values. Their enums opt in, so the recordings still match what the code sends.
- Recording the Gemini test found the reviewer's worry to be real: with the folded `anyOf` of one-value `enum`s, `gemini-2.5-flash` called the tool with `priority='Urgent'`, a value outside the options, on one of two recordings. `GoogleJsonSchemaTransformer` now folds such an `anyOf` back into one `enum` and moves each option's description into the parent's (`How urgent the ticket is.\nlow: Can wait a week.\nhigh: Needs attention today.`), so Gemini gets the shape it holds the model to. `tests/test_json_schema.py::test_described_options_fold_into_one_enum_google` pins the fold and that a union of real alternatives is left alone.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] Any permitted **compatibility impact** has a label, warning, migration, release note, and exact API-check waiver.
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

### CODE REVIEW

**Comment by DouweM** [comment]
Measured, since the recorded `Urgent` could have been a one-off. `gemini-2.5-flash`, a `set_priority(priority)` tool, 20 runs per shape over four tickets, counting the value the model called with:

| Shape on the wire | in the enum | outside the enum | no call |
|---|---|---|---|
| `anyOf` of one-value `enum`s with descriptions (before the fold) | 1 | 8 (`Urgent` ×6, `Highest`, `1 - Critical`) | 11 |
| plain `enum`, no docstrings (today's behaviour) | 12 | 0 | 8 |
| one `enum`, descriptions folded into the parent's (this PR) | 13 | 0 | 7 |

So the `anyOf` shape does lose Gemini's hold on the options almost entirely, and the fold restores it while keeping the descriptions. Script: `gemini_enum_adherence.py` in the session scratchpad; the "no call" rows are the model answering in prose on the low-priority tickets, the same across shapes.

File: pydantic_ai_slim/pydantic_ai/_utils.py [hunk 1]
```diff
@@ -1129,3 +1132,30 @@ def estimate_string_tokens(text: str) -> int:
...
+        return {}
+    docstrings: dict[str, str] = {}
+    for previous, node in zip(class_def.body, class_def.body[1:]):
+        if not (
+            isinstance(node, ast.Expr) and isinstance(node.value, ast.Constant) and isinstance(node.value.value, str)
+        ):
+            continue
+        if not isinstance(previous, ast.Assign):
+            continue
+        for name in [target.id for target in previous.targets if isinstance(target, ast.Name)]:
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/_utils.py]: Right: a string after `_ignore_` (or any non-member name) is not an option's description. Names are now filtered through `cls.__members__` in 5b6ffa15c, with `_ignore_` in the test enum.

[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/_utils.py]: Annotated members (`ast.AnnAssign`) are read too as of 5b6ffa15c; the test enum has one.

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 2]
```diff
@@ -271,6 +271,18 @@ async def matches_tool_selector(
 
 
 class GenerateToolJsonSchema(GenerateJsonSchema):
+    def enum_schema(self, schema: core_schema.EnumSchema) -> JsonSchemaValue:
+        # A docstring under an enum member describes that option, as `anyOf` of `const`s with descriptions
+        # (the JSON Schema way to describe single values), so models can tell the options apart.
+        json_schema = super().enum_schema(schema)
+        docstrings = _utils.enum_member_docstrings(schema['cls'])
+        if docstrings:
+            json_schema['anyOf'] = [
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/tools.py:305]: Confirmed: `_get_python_type` took the first `anyOf` branch and made it `Optional[Any]`. As of 5b6ffa15c an `anyOf` of `const`s maps to the type of its values (the parent's, or inferred), so a documented enum is `str` in the prompted format again; unit case in `test_generate_user_output_format_complex`, and a recorded Mistral tool call with a documented enum in be802e499.

File: pydantic_ai_slim/pydantic_ai/profiles/google.py [hunk 3]
```diff
@@ -400,3 +401,32 @@ def transform(self, schema: JsonSchema) -> JsonSchema:
...
+    if not options or not all(
+        isinstance(option, dict)
+        and len(option.get('enum', ())) == 1
+        and option.keys() <= {'enum', 'type', 'description'}
+        for option in options
+    ):
+        return
+    types = {option.get('type') for option in options}
+    if len(types) != 1:
+        return
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/profiles/google.py:427]: Agreed. As of 5b6ffa15c the fold only applies when the parent has no `enum` of its own and its `type`, if any, is the options' type; anything else is left as the `anyOf` Gemini already accepted. Both cases are pinned in `test_described_options_fold_into_one_enum_google`.

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 4]
```diff
@@ -271,6 +271,18 @@ async def matches_tool_selector(
...
+                {'const': value, **({'description': docstrings[member.name]} if member.name in docstrings else {})}
+                for member, value in zip(schema['members'], json_schema.pop('enum'))
+            ]
+        return json_schema
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/tools.py:309]: Scoped as: fold back where a provider does not hold the model to the shape (Gemini, measured on the PR), map the shape to its plain type where a provider derives a Python type from it (Mistral's prompted format, fixed in 5b6ffa15c), and record the rest. be802e499 adds recorded tool calls with a documented enum for OpenAI (Responses, strict schema, the `anyOf` of `const`s goes through untouched and the call comes back inside the enum) and Mistral, beside the Anthropic and Gemini ones. Bedrock and Cohere are not recorded; their transformers do not special-case `anyOf`, so they see what Anthropic and OpenAI see.

**Comment by DouweM** [comment]
Update: #8450 no longer stacks on this branch. It stands on `main` again with the enum change reversed out, so the Jev release does not wait on the enum decision; Jev's `IntEnum` rubrics and described options come back as a follow-up on top of this PR.

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 5]
```diff
@@ -271,6 +271,24 @@ async def matches_tool_selector(
...
 class GenerateToolJsonSchema(GenerateJsonSchema):
+    describe_enum_members: ClassVar[bool] = False
+    """Read the docstring under each `Enum` member even when the enclosing model does not opt in."""
+
+    def enum_schema(self, schema: core_schema.EnumSchema) -> JsonSchemaValue:
+        # A docstring under an enum member describes that option, as `anyOf` of `const`s with descriptions
+        # (the JSON Schema way to describe single values), so models can tell the options apart. Read on the
+        # same switch as a docstring under a field, `use_attribute_docstrings` on the enclosing model, which a
+        # tool's parameters set; a model of the user's own opts in, and a bare output is described for it.
+        json_schema = super().enum_schema(schema)
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/tools.py:285]: Confirmed and fixed in `1fa3b4430`, though the symptom is different from the one predicted.

The Google transformer does not reject the request: it strips the unrecognised `const: null` and leaves the branch as `{'description': ...}` with no `type` and no `enum`, which matches any value. So a documented enum with a `None` member silently *loses* its constraint rather than failing loudly. Reproduced on this branch before the fix:

```
before: anyOf: [{const: 'yes', description: ...}, {const: null, description: ...}]
after Google: anyOf: [{description: ..., enum: ['yes'], type: 'string'}, {description: ...}]
```

The fix is at the source rather than in the transformer. `enum_schema` now only produces the described `anyOf` when every member value is non-null; an enum carrying a `None` keeps the plain `enum` list, which states every value including the null and is exactly what such an enum produced before this PR. That keeps the change to what it is for, describing string options, and leaves no way for this PR to emit a `const: null` at all.

Regression test: `test_a_none_member_keeps_the_enum_in_its_plain_form`.

**Comment by DouweM** [comment]
Re-requesting review at the current head. The `CHANGES_REQUESTED` above was left at `752829897`; since then every finding on this PR has been addressed and every thread resolved, most recently the `None`-member enum in `1fa3b4430`, and CI is green on 69 checks.

File: pydantic_ai_slim/pydantic_ai/tools.py [hunk 6]
```diff
@@ -271,6 +271,27 @@ async def matches_tool_selector(
...
+        read = self.describe_enum_members or self._config.use_attribute_docstrings
```
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/tools.py]: Confirmed, and it is a regression against this PR's own documented rule rather than a disagreement about intent. Reproduced on the branch head:

```python
class Priority(str, Enum):
    """How urgent the ticket is."""
    low = 'low'
    """Can wait a week."""
    high = 'high'
    """Needs attention today."""

@agent.tool_plain
def triage(priority: Priority) -> str: ...
```

The schema the model receives:

```json
{"Priority": {"description": "How urgent the ticket is.", "enum": ["low", "high"], "title": "Priority", "type": "string"}}
```

The member descriptions are gone. `docs/tools.md` says "Tool parameters always read them", and all four recorded cassettes still contain `Can wait a week`, so the wire tests were recorded while this worked and the snapshots were later updated to match the broken output. Your sibling comments on `docs/tools.md` and `tests/models/test_mistral.py` are the same defect seen from two other angles.

**Why it is off.** `enum_schema` reads `self.describe_enum_members or self._config.use_attribute_docstrings`. For an enum inside a user's own model the config is that model's, so the opt-in works. For a function tool the schema comes from the synthetic `TypedDict` that `_function_schema` builds, whose config has `use_attribute_docstrings` at its default of `False`, so neither half of the condition is true. Only `_GenerateBareOutputJsonSchema` sets the ClassVar.

**Why I have not just fixed it.** The obvious one-line change, setting `describe_enum_members = True` on `GenerateToolJsonSchema`, also un-gates output models, because an output `BaseModel` uses that same generator and is currently gated through `_config`. That would undo the deliberate decision to read member docstrings only on Pydantic's own `use_attribute_docstrings` switch.

So it needs a decision, and it is the maintainer's:

1. **A separate generator for function-tool parameters**, so a tool's parameters always describe and a user's own model stays gated. Matches the documented rule, but a single model-like argument is a tool's parameters *and* a user's model, so one of the two rules has to win for it.
2. **Gate tool parameters too**, and change the docs sentence instead. Simplest, and consistent, but then the feature does nothing for the case the PR leads with.
3. **Set `use_attribute_docstrings=True` on the synthetic `TypedDict`**, which also makes field docstrings readable there. Wider than this feature.

Leaving this thread open rather than resolving it, since the PR should not merge in this state either way.

File: docs/tools.md [hunk 7]
```diff
@@ -266,6 +266,12 @@ first entry of the returns section. Other sections Griffe can parse, such as `Ra
 `Warnings` and `Yields`, are dropped, so anything the model needs to act on belongs in the
 leading description, a parameter description, or the first returns entry.
 
+A docstring under an `Enum` member reaches the model too, as the description of that value, so the model sees
+what each option means and not only its name. Such an enum renders as `anyOf` of `const` values instead of a
+plain `enum` list; members without docstrings, and `Literal`s, render as before. Tool parameters always read
```
[the remark this answers is not in the document]
Comment (DouweM) [reply docs/tools.md]: Confirmed, and it is a regression against this PR's own documented rule rather than a disagreement about intent. Reproduced on the branch head:

```python
class Priority(str, Enum):
    """How urgent the ticket is."""
    low = 'low'
    """Can wait a week."""
    high = 'high'
    """Needs attention today."""

@agent.tool_plain
def triage(priority: Priority) -> str: ...
```

The schema the model receives:

```json
{"Priority": {"description": "How urgent the ticket is.", "enum": ["low", "high"], "title": "Priority", "type": "string"}}
```

The member descriptions are gone. `docs/tools.md` says "Tool parameters always read them", and all four recorded cassettes still contain `Can wait a week`, so the wire tests were recorded while this worked and the snapshots were later updated to match the broken output. Your sibling comments on `docs/tools.md` and `tests/models/test_mistral.py` are the same defect seen from two other angles.

**Why it is off.** `enum_schema` reads `self.describe_enum_members or self._config.use_attribute_docstrings`. For an enum inside a user's own model the config is that model's, so the opt-in works. For a function tool the schema comes from the synthetic `TypedDict` that `_function_schema` builds, whose config has `use_attribute_docstrings` at its default of `False`, so neither half of the condition is true. Only `_GenerateBareOutputJsonSchema` sets the ClassVar.

**Why I have not just fixed it.** The obvious one-line change, setting `describe_enum_members = True` on `GenerateToolJsonSchema`, also un-gates output models, because an output `BaseModel` uses that same generator and is currently gated through `_config`. That would undo the deliberate decision to read member docstrings only on Pydantic's own `use_attribute_docstrings` switch.

So it needs a decision, and it is the maintainer's:

1. **A separate generator for function-tool parameters**, so a tool's parameters always describe and a user's own model stays gated. Matches the documented rule, but a single model-like argument is a tool's parameters *and* a user's model, so one of the two rules has to win for it.
2. **Gate tool parameters too**, and change the docs sentence instead. Simplest, and consistent, but then the feature does nothing for the case the PR leads with.
3. **Set `use_attribute_docstrings=True` on the synthetic `TypedDict`**, which also makes field docstrings readable there. Wider than this feature.

Leaving this thread open rather than resolving it, since the PR should not merge in this state either way.

File: tests/models/test_mistral.py [hunk 8]
```diff
@@ -3709,3 +3718,54 @@ async def test_parallel_tool_calls_stream(allow_model_requests: None) -> None:
...
+    """Can wait a week."""
+    high = 'high'
+    """Needs attention today."""
+
+
+@pytest.mark.vcr()
+async def test_mistral_enum_member_docstrings_reach_the_wire(
+    allow_model_requests: None, mistral_api_key: str, request_capture: RequestCapture
+):
+    """A documented enum goes to Mistral as `anyOf` of `const`s with descriptions, and the model calls with one."""
```
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/models/test_mistral.py:3758]: Confirmed, and it is a regression against this PR's own documented rule rather than a disagreement about intent. Reproduced on the branch head:

```python
class Priority(str, Enum):
    """How urgent the ticket is."""
    low = 'low'
    """Can wait a week."""
    high = 'high'
    """Needs attention today."""

@agent.tool_plain
def triage(priority: Priority) -> str: ...
```

The schema the model receives:

```json
{"Priority": {"description": "How urgent the ticket is.", "enum": ["low", "high"], "title": "Priority", "type": "string"}}
```

The member descriptions are gone. `docs/tools.md` says "Tool parameters always read them", and all four recorded cassettes still contain `Can wait a week`, so the wire tests were recorded while this worked and the snapshots were later updated to match the broken output. Your sibling comments on `docs/tools.md` and `tests/models/test_mistral.py` are the same defect seen from two other angles.

**Why it is off.** `enum_schema` reads `self.describe_enum_members or self._config.use_attribute_docstrings`. For an enum inside a user's own model the config is that model's, so the opt-in works. For a function tool the schema comes from the synthetic `TypedDict` that `_function_schema` builds, whose config has `use_attribute_docstrings` at its default of `False`, so neither half of the condition is true. Only `_GenerateBareOutputJsonSchema` sets the ClassVar.

**Why I have not just fixed it.** The obvious one-line change, setting `describe_enum_members = True` on `GenerateToolJsonSchema`, also un-gates output models, because an output `BaseModel` uses that same generator and is currently gated through `_config`. That would undo the deliberate decision to read member docstrings only on Pydantic's own `use_attribute_docstrings` switch.

So it needs a decision, and it is the maintainer's:

1. **A separate generator for function-tool parameters**, so a tool's parameters always describe and a user's own model stays gated. Matches the documented rule, but a single model-like argument is a tool's parameters *and* a user's model, so one of the two rules has to win for it.
2. **Gate tool parameters too**, and change the docs sentence instead. Simplest, and consistent, but then the feature does nothing for the case the PR leads with.
3. **Set `use_attribute_docstrings=True` on the synthetic `TypedDict`**, which also makes field docstrings readable there. Wider than this feature.

Leaving this thread open rather than resolving it, since the PR should not merge in this state either way.

**Comment by DouweM** [comment]
Not ready to merge. The re-run review found something real that the earlier rounds missed, and I have left the three threads open.

**The feature does not work for tool parameters**, which is the case this PR leads with. A documented `Enum` used as a `@agent.tool_plain` parameter reaches the model as a plain `enum` with only the class docstring; the member descriptions are dropped. `docs/tools.md` says "Tool parameters always read them", and all four recorded cassettes still contain `Can wait a week`, so this worked when they were recorded and regressed afterwards, with the snapshots updated to match the broken output.

It happens because `enum_schema` reads `self.describe_enum_members or self._config.use_attribute_docstrings`, and a function tool's schema is built from a synthetic `TypedDict` whose config has that switch off.

I have not fixed it because the fix is a design choice: the one-line version, setting the ClassVar on `GenerateToolJsonSchema`, would also un-gate output models, which is the opposite of the decision to read member docstrings only on Pydantic's own switch. The three options are laid out on the first thread.

Everything else on this PR is done: CI is green on 68 checks, and every earlier finding is addressed, most recently the `None`-member enum in `1fa3b4430` that silently dropped an option's constraint.

File: pydantic_ai_slim/pydantic_ai/tools.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DouweM) [reply pydantic_ai_slim/pydantic_ai/tools.py]: Right on every point, and thank you for chasing it to the cassettes — that is the part I had not checked.

Douwe's call on your question: **tool parameters are described.** The gate is gone in f90ac9b75. `self._config` is pushed while the *core* schema is built and nothing pushes it while the JSON schema is generated, so an enum reached from a tool's synthetic model never saw the `use_attribute_docstrings=True` that `_function_schema` already sets — the intent was declared and simply did not arrive. `enum_schema` now reads the docstrings unconditionally, which is one rule instead of a rule that held in three places out of four with nothing to tell a user which one they were in.

On the recordings: all four cassettes still hold the `anyOf` with `Can wait a week.`/`Needs attention today.`, and restoring the behaviour makes the live request that shape again, so the code and the recordings agree without re-recording — the fix closes the drift from the other side. The snapshots are the real guard here anyway, since they assert the body `request_capture` saw rather than the replayed one, and they now assert `anyOf` on all four providers.

Two things moved with it that were not in the PR before: `test_enum_member_docstrings_follow_the_models_docstring_switch` is renamed and rewritten, because the switch it named no longer exists; and `TypeSafeModel` gained option descriptions it was silently dropping — its Choice `criteria` were going out as `{'ask': None, 'reject': None, 'run': None}` with the docstrings written but never sent.

File: docs/tools.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (DouweM) [reply docs/tools.md]: It holds now. The gate is removed in f90ac9b75, so a tool parameter's enum is described like any other, and the four wire snapshots assert the `anyOf` form read live from `request_capture`. No change needed to this sentence.

File: tests/models/test_mistral.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DouweM) [reply tests/models/test_mistral.py:3758]: The docstring was right and the assertion was wrong; f90ac9b75 makes the code match both. The snapshot now asserts the `anyOf` of `const`s the docstring describes, on this test and on the OpenAI, Anthropic and Google ones, and the cassettes already held that shape — so the recording and the live request agree again rather than needing to be re-recorded.

**Comment by DouweM** [comment]
Filed #8522 for the follow-up: the `Docstrings` section this PR adds documents behaviour that is accidental rather than designed — a `dataclass` describes its fields as a tool parameter and stops doing so as an output type, and a `pydantic.dataclasses.dataclass` describes less than a plain one. That section should shrink to a sentence once #8522 lands.
